app/api/search/search.js

```diff
@@ -284,57 +293,154 @@ const _getAggregationDictionary = async (aggregation, language, property, dictio
+      labeledBuckets.push(newBucket);
+    }
+  });
+
+  return labeledBuckets;
+}
+
+const SIMPLE_LIMIT_KEYS = new Set(['_types', 'generatedToc', '_permissions.self', '_published']);
+const PERMISSION_KEYS = new Set(['_permissions.read', '_permissions.write']);
+const _denormalizeAndLimitAggregations = async (
+  aggregations,
+  templates,
+  dictionaries,
+  language,
+  limit
+) => {
   const properties = propertiesHelper.allProperties(templates);
+  const propertiesByName = objectIndex(
+    properties,
+    p => p.name,
+    p => p
+  );
+  const propertiesById = objectIndex(
+    properties,
+    p => p._id.toString(),
+    p => p
+  );
   const newRelationshipsEnabled = v2.checkFeatureEnabled();
-  return Object.keys(aggregations).reduce(async (denormaLizedAgregationsPromise, key) => {
-    const denormaLizedAgregations = await denormaLizedAgregationsPromise;
-    if (
-      !aggregations[key].buckets ||
-      aggregations[key].type === 'nested' ||
-      ['_types', 'generatedToc', '_permissions.self', '_published'].includes(key)
-    ) {
-      return Object.assign(denormaLizedAgregations, { [key]: aggregations[key] });
+  const dictionariesById = objectIndex(
+    dictionaries,
+    d => d._id.toString(),
+    d => d
+  );
+  const dictionaryCache = {};
+  const denormalizedAggregations = {};
+  // eslint-disable-next-line max-statements
+  await sequentialPromises(Object.keys(aggregations), async key => {
+    if (!aggregations[key].buckets || aggregations[key].type === 'nested') {
+      denormalizedAggregations[key] = aggregations[key];
+      return;
+    }
+
+    if (SIMPLE_LIMIT_KEYS.has(key)) {
+      denormalizedAggregations[key] = {
+        ...aggregations[key],
+        buckets: limitBuckets(aggregations[key].buckets, limit),
+      };
+      return;
     }
 
-    if (['_permissions.read', '_permissions.write'].includes(key)) {
+    if (PERMISSION_KEYS.has(key)) {
       const [users, groups] = await Promise.all([usersModel.get(), userGroups.get()]);
 
       const info = [
         ...users.map(u => ({ type: 'user', refId: u._id, label: u.username })),
         ...groups.map(g => ({ type: 'group', refId: g._id, label: g.name })),
       ];
+      const infoByRefId = objectIndex(
+        info,
+        i => i.refId.toString(),
+        i => i
+      );
 
       const role = permissionsContext.getUserInContext()?.role;
       const refIds = permissionsContext.permissionsRefIds();
 
-      const buckets = aggregations[key].buckets
-        .filter(bucket => role === UserRole.ADMIN || refIds.includes(bucket.key))
+      let buckets = limitBuckets(aggregations[key].buckets, limit);
+      buckets =
+        role === UserRole.ADMIN ? buckets : buckets.filter(bucket => refIds.includes(bucket.key));
+      buckets = buckets
         .map(bucket => {
-          const itemInfo = info.find(i => i.refId.toString() === bucket.key);
+          const itemInfo = bucket.key in infoByRefId ? infoByRefId[bucket.key] : null;
```

Comment:
If I understand the purpose of this check correctly, it should have the same result as just `const itemInfo = infoByRefId?.[bucket.key]` since the possible `null` doesn't add to the algorithm.